cpp/src/interpolate/cubic_spline.cu

```diff
@@ -0,0 +1,404 @@
+                  cudf::column_view const& prefixes,
+                  cudf::column_view const& query_coords,
+                  cudf::mutable_column_view const& result,
+                  rmm::mr::device_memory_resource *mr,
+                  cudaStream_t stream) {
+      const T* T_ = t.data<T>();
+      const int32_t* CURVE_IDS_ = curve_ids.data<int32_t>();
+      const int32_t* PREFIXES_ = prefixes.data<int32_t>();
+      const T* QUERY_COORDS_ = query_coords.data<T>();
+      T* RESULT_ = result.data<T>();
+      thrust::for_each(rmm::exec_policy(stream)->on(stream),
+        thrust::make_counting_iterator<int>(0),
+        thrust::make_counting_iterator<int>(query_coords.size()),
+        [T_, CURVE_IDS_, PREFIXES_, QUERY_COORDS_, RESULT_] __device__
+        (int index) {
+          int curve = CURVE_IDS_[index];
+          int len = PREFIXES_[curve+1] - PREFIXES_[curve];
+          int h = PREFIXES_[curve];
+          int dh = PREFIXES_[curve] - (curve);
+          // O(n) search, can do log(n) easily
+          for(int32_t i = 0 ; i < len ; ++i) {
+            if((T_[h+i]+0.0001 - QUERY_COORDS_[index]) > 0.00001) {
+              RESULT_[index] = dh+i-1;
+              if(i == 0) RESULT_[index] = index-curve;
+              return;
+            }
+          }
+          // TODO: Important failure case:
+          // This will use the final set of coefficients
+          // for t_ values that are outside of the original
+          // interpolation range.
+          RESULT_[index] = h+len - 2;
+      });
+  };
+  template <typename T>
+  std::enable_if_t<not std::is_floating_point<T>::value, void>
+  operator()(cudf::column_view const& t,
+                  cudf::column_view const& curve_ids,
+                  cudf::column_view const& prefixes,
+                  cudf::column_view const& query_coords,
+                  cudf::mutable_column_view const& result,
+                  rmm::mr::device_memory_resource *mr,
+                  cudaStream_t stream) {
+      CUDF_FAIL("Non-floating point operation is not supported.");
+  }
+};
+
+struct interpolate {
+  template <typename T>
+  std::enable_if_t<std::is_floating_point<T>::value, void>
+  operator()(cudf::column_view const& t,
+                  cudf::column_view const& ids,
+                  cudf::column_view const& coef_indexes,
+                  cudf::table_view const& coefficients,
+                  cudf::mutable_column_view const& result,
+                  rmm::mr::device_memory_resource *mr,
+                  cudaStream_t stream) {
+      const T* T_ = t.data<T>();
+      const int32_t* IDS_ = ids.data<int32_t>();
+      const int32_t* COEF_INDEXES_ = coef_indexes.data<int32_t>();
+      const T* D3_ = coefficients.column(3).data<T>();
+      const T* D2_ = coefficients.column(2).data<T>();
+      const T* D1_ = coefficients.column(1).data<T>();
+      const T* D0_ = coefficients.column(0).data<T>();
+      T* RESULT_ = result.data<T>();
```

Comment:
Create the output column here, rather than taking a mutable column view.
```suggestion
      auto result = make_numeric_column(...)
```

Reply:
Here's where I can't do that: I'd need `offsets.size()` number of return columns which would be allocated inside of the functor, then I'd have to concatenate them into a single long SoA result column before returning the result.

Reply:
Hmmm, I don't see that.  the for_each knows how many elements are in the output, so that size is definitely available. And the for_each is only writing to one result array. So what I suggest is doable.